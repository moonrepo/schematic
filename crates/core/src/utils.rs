use proc_macro2::TokenStream;
use quote::quote;
use syn::{Attribute, Expr, Meta, Path};

pub fn get_meta_path(meta: &Meta) -> &Path {
    match meta {
        Meta::Path(path) => path,
        Meta::List(list) => &list.path,
        Meta::NameValue(nv) => &nv.path,
    }
}

pub fn preserve_str_literal(meta: &Meta) -> darling::Result<Expr> {
    match meta {
        Meta::Path(_) => Err(darling::Error::unsupported_format("path").with_span(meta)),
        Meta::List(_) => Err(darling::Error::unsupported_format("list").with_span(meta)),
        Meta::NameValue(nv) => Ok(nv.value.clone()),
    }
}

pub fn is_inheritable_attribute(attr: &Attribute) -> bool {
    let path = get_meta_path(&attr.meta);

    ["allow", "default", "deprecated", "doc", "warn"]
        .into_iter()
        .any(|n| path.is_ident(n))
}

pub fn to_type_string(ts: TokenStream) -> String {
    format!("{ts}")
        .replace(" :: ", "::")
        .replace(" , ", ", ")
        .replace(" < ", "<")
        .replace("< ", "<")
        .replace(" <", "<")
        .replace(" > ", ">")
        .replace("> ", ">")
        .replace(" >", ">")
}

#[derive(Default)]
pub struct ImplResult {
    pub requires_internal: bool,
    pub no_value: bool,
    pub value: TokenStream,
}

impl ImplResult {
    pub fn skipped() -> Self {
        Self {
            no_value: true,
            ..Default::default()
        }
    }

    pub fn impl_struct_default(show: bool) -> TokenStream {
        if show {
            quote! { ..Default::default() }
        } else {
            quote! {}
        }
    }

    pub fn impl_use_internal(show: bool) -> TokenStream {
        if show {
            quote! { use schematic::internal::*; }
        } else {
            quote! {}
        }
    }
}
