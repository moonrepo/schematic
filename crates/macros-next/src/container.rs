use crate::args::{SerdeContainerArgs, SerdeRenameArg};
use crate::field::Field;
use crate::variant::Variant;
use darling::FromDeriveInput;
use std::rc::Rc;
use syn::{Attribute, Data, DeriveInput, ExprPath, Fields, Ident, Visibility};

// #[config()], #[schematic()]
#[derive(Default, FromDeriveInput)]
#[darling(
    default,
    attributes(config, schematic),
    supports(struct_named, enum_any)
)]
pub struct ContainerArgs {
    // config
    pub allow_unknown_fields: bool,
    pub context: Option<ExprPath>,
    // pub partial: PartialAttr, // TODO
    #[cfg(feature = "env")]
    pub env_prefix: Option<String>,

    // serde
    pub rename: Option<SerdeRenameArg>,
    pub rename_all: Option<SerdeRenameArg>,
    pub rename_all_fields: Option<SerdeRenameArg>,
    // pub serde: SerdeMeta, // TODO
}

pub struct Container {
    pub args: Rc<ContainerArgs>,
    pub inner: ContainerInner,
    pub serde_args: Rc<SerdeContainerArgs>,

    // inherited
    pub attrs: Vec<Attribute>,
    pub ident: Ident,
    pub vis: Visibility,
}

impl Container {
    pub fn from(input: DeriveInput) -> Self {
        let args = Rc::new(ContainerArgs::from_derive_input(&input).unwrap_or_default());
        let serde_args = Rc::new(SerdeContainerArgs::from_derive_input(&input).unwrap_or_default());

        let inner = match input.data {
            Data::Struct(data) => match data.fields {
                Fields::Named(fields) => ContainerInner::NamedStruct {
                    fields: fields
                        .named
                        .into_iter()
                        .map(|data| Field::new(data, args.clone(), serde_args.clone()))
                        .collect(),
                },
                Fields::Unnamed(fields) => ContainerInner::UnnamedStruct {
                    fields: fields
                        .unnamed
                        .into_iter()
                        .enumerate()
                        .map(|(index, data)| {
                            let mut field = Field::new(data, args.clone(), serde_args.clone());
                            field.index = index;
                            field
                        })
                        .collect(),
                },
                Fields::Unit => {
                    panic!("Unit structs are not supported.");
                }
            },
            Data::Enum(data) => {
                let all_unit = data
                    .variants
                    .iter()
                    .all(|variant| matches!(variant.fields, Fields::Unit));
                let variants = data
                    .variants
                    .into_iter()
                    .map(|data| Variant::new(data, args.clone(), serde_args.clone()))
                    .collect::<Vec<_>>();

                if all_unit {
                    ContainerInner::UnitEnum { variants }
                } else {
                    ContainerInner::Enum { variants }
                }
            }
            Data::Union(_) => {
                panic!("Unions are not supported.");
            }
        };

        Self {
            args,
            attrs: input.attrs,
            ident: input.ident,
            inner,
            serde_args,
            vis: input.vis,
        }
    }
}

pub enum ContainerInner {
    NamedStruct { fields: Vec<Field> },
    UnnamedStruct { fields: Vec<Field> },
    Enum { variants: Vec<Variant> },
    UnitEnum { variants: Vec<Variant> },
}
