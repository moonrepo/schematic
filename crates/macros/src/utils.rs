use convert_case::{Case, Casing};
use syn::{AngleBracketedGenericArguments, Attribute, Expr, Meta};

pub fn format_case(format: &str, value: &str) -> String {
    let case = match format {
        "lowercase" => Case::Lower,
        "UPPERCASE" => Case::Upper,
        "PascalCase" => Case::Pascal,
        "camelCase" => Case::Camel,
        "snake_case" => Case::Snake,
        "SCREAMING_SNAKE_CASE" => Case::UpperSnake,
        "SCREAMING-KEBAB-CASE" => Case::UpperKebab,
        _ => Case::Kebab,
    };

    value.to_case(case)
}

pub fn preserve_str_literal(meta: &Meta) -> darling::Result<Expr> {
    match meta {
        Meta::Path(_) => Err(darling::Error::unsupported_format("path").with_span(meta)),
        Meta::List(_) => Err(darling::Error::unsupported_format("list").with_span(meta)),
        Meta::NameValue(nv) => Ok(nv.value.clone()),
    }
}

pub fn extract_common_attrs(attrs: &[Attribute]) -> Vec<&Attribute> {
    let preserve = ["allow", "deprecated", "doc", "warn"];

    attrs
        .iter()
        .filter(|a| {
            let path = match &a.meta {
                Meta::Path(path) => path,
                Meta::List(list) => &list.path,
                Meta::NameValue(nv) => &nv.path,
            };

            preserve.iter().any(|n| path.is_ident(n))
        })
        .collect()
}

// Thanks to confique for the implementation:
// https://github.com/LukasKalbertodt/confique/blob/main/macro/src/util.rs
pub fn unwrap_path_type<'l>(
    ty: &'l syn::Type,
    lookups: &[&[&str]],
) -> Option<&'l AngleBracketedGenericArguments> {
    let ty = match ty {
        syn::Type::Path(path) => path,
        _ => return None,
    };

    if ty.qself.is_some() || ty.path.leading_colon.is_some() {
        return None;
    }

    if !lookups
        .iter()
        .any(|vp| ty.path.segments.iter().map(|s| &s.ident).eq(*vp))
    {
        return None;
    }

    match &ty.path.segments.last().unwrap().arguments {
        syn::PathArguments::AngleBracketed(args) => Some(args),
        _ => None,
    }
}

pub fn unwrap_option(ty: &syn::Type) -> Option<&syn::Type> {
    let Some(args) = unwrap_path_type(
        ty,
        &[
            &["Option"],
            &["std", "option", "Option"],
            &["core", "option", "Option"],
        ],
    ) else {
        return None;
    };

    if args.args.len() != 1 {
        return None;
    }

    match &args.args[0] {
        syn::GenericArgument::Type(t) => Some(t),
        _ => None,
    }
}
