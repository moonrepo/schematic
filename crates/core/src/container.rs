use crate::args::{PartialArg, SerdeContainerArgs, SerdeRenameArg};
use crate::field::Field;
use crate::utils::{ImplResult, is_inheritable_attribute};
use crate::variant::Variant;
use darling::FromDeriveInput;
use proc_macro2::TokenStream;
use quote::{ToTokens, format_ident, quote};
use std::rc::Rc;
use syn::{Attribute, Data, DeriveInput, ExprPath, Fields, Ident, Visibility};

// #[config()], #[schematic()]
#[derive(Debug, Default, FromDeriveInput)]
#[darling(default, attributes(config, schematic), supports(struct_any, enum_any))]
pub struct ContainerArgs {
    // config
    pub allow_unknown_fields: bool,
    pub context: Option<ExprPath>,
    #[cfg(feature = "env")]
    pub env_prefix: Option<String>,
    pub partial: Option<PartialArg>,

    // serde
    pub rename: Option<SerdeRenameArg>,
    pub rename_all: Option<SerdeRenameArg>,
    pub rename_all_fields: Option<SerdeRenameArg>,
}

#[derive(Debug)]
pub struct Container {
    pub args: Rc<ContainerArgs>,
    pub inner: ContainerInner,
    pub serde_args: Rc<SerdeContainerArgs>,

    // inherited
    pub attrs: Vec<Attribute>,
    pub ident: Ident,
    pub vis: Visibility,
}

impl Container {
    pub fn from(input: DeriveInput) -> Self {
        let args = Rc::new(ContainerArgs::from_derive_input(&input).unwrap());
        let serde_args = Rc::new(SerdeContainerArgs::from_derive_input(&input).unwrap());

        let inner = match input.data {
            Data::Struct(data) => match data.fields {
                Fields::Named(fields) => ContainerInner::NamedStruct {
                    fields: fields
                        .named
                        .into_iter()
                        .map(|data| Field::new(data, args.clone(), serde_args.clone()))
                        .collect(),
                },
                Fields::Unnamed(fields) => ContainerInner::UnnamedStruct {
                    fields: fields
                        .unnamed
                        .into_iter()
                        .enumerate()
                        .map(|(index, data)| {
                            let mut field = Field::new(data, args.clone(), serde_args.clone());
                            field.index = index;
                            field
                        })
                        .collect(),
                },
                Fields::Unit => {
                    panic!("Unit structs are not supported.");
                }
            },
            Data::Enum(data) => {
                let all_unit = data
                    .variants
                    .iter()
                    .all(|variant| matches!(variant.fields, Fields::Unit));
                let variants = data
                    .variants
                    .into_iter()
                    .map(|data| Variant::new(data, args.clone(), serde_args.clone()))
                    .collect::<Vec<_>>();

                if all_unit {
                    ContainerInner::UnitEnum { variants }
                } else {
                    ContainerInner::Enum { variants }
                }
            }
            Data::Union(_) => {
                panic!("Unions are not supported.");
            }
        };

        let container = Self {
            args,
            attrs: input.attrs,
            ident: input.ident,
            inner,
            serde_args,
            vis: input.vis,
        };
        container.validate_args();
        container
    }

    fn validate_args(&self) {}

    pub fn get_partial_attributes(&self) -> Vec<TokenStream> {
        let serde_args = self.get_partial_serde_attribute_args();
        let mut attrs = vec![quote! { #[serde(#serde_args) ]}];

        for attr in &self.attrs {
            if is_inheritable_attribute(attr) {
                attrs.push(quote! { #attr });
            }
        }

        // TODO
        // let partial = &self.args.partial;
        // attrs.push(quote! { #partial });

        attrs
    }

    pub fn get_partial_serde_attribute_args(&self) -> TokenStream {
        let mut meta = vec![];

        match &self.inner {
            ContainerInner::NamedStruct { .. } => {
                meta.push(quote! { default });

                if self.serde_args.deny_unknown_fields || !self.args.allow_unknown_fields {
                    meta.push(quote! { deny_unknown_fields });
                }
            }
            ContainerInner::UnnamedStruct { .. } => {
                meta.push(quote! { default });
            }
            ContainerInner::Enum { .. } => {
                if let Some(tag) = &self.serde_args.tag {
                    meta.push(quote! { tag = #tag });
                }

                if let Some(content) = &self.serde_args.content {
                    meta.push(quote! { content = #content });
                }

                if self.serde_args.untagged {
                    meta.push(quote! { untagged });
                }
            }
            ContainerInner::UnitEnum { .. } => {
                meta.push(quote! { untagged });
            }
        };

        if let Some(expecting) = &self.serde_args.expecting {
            meta.push(quote! { expecting = #expecting });
        }

        if let Some(rename) = &self.serde_args.rename {
            meta.push(rename.get_meta("rename"));
        }

        if let Some(rename_all) = &self.serde_args.rename_all {
            meta.push(rename_all.get_meta("rename_all"));
        }

        if let Some(rename_all_fields) = &self.serde_args.rename_all_fields {
            meta.push(rename_all_fields.get_meta("rename_all_fields"));
        }

        quote! {
            #(#meta),*
        }
    }

    pub fn impl_partial(&self) -> TokenStream {
        let base_name = &self.ident;
        let partial_name = format_ident!("Partial{base_name}");
        let context = match self.args.context.as_ref() {
            Some(ctx) => quote! { #ctx },
            None => quote! { () },
        };

        let default_values_method = self.impl_partial_default_values();
        let env_values_method = self.impl_partial_env_values();
        let extends_from_method = self.impl_partial_extends_from();
        let merge_method = self.impl_partial_merge();

        quote! {
            #[automatically_derived]
            impl schematic::PartialConfig for #partial_name {
                type Context = #context;

                #default_values_method
                #env_values_method
                #extends_from_method
                #merge_method
            }

            #[automatically_derived]
            impl schematic::Config for #base_name {
                // TODO
            }

            #[automatically_derived]
            impl Default for #base_name {
                fn default() -> Self {
                    <Self as schematic::Config>::from_partial(
                        <Self as schematic::Config>::default_partial()
                    )
                }
            }
        }
    }

    pub fn impl_partial_default_values(&self) -> TokenStream {
        let mut requires_internal = false;

        let inner = match &self.inner {
            ContainerInner::NamedStruct { fields } => {
                let mut rows = vec![];

                for field in fields {
                    let res = field.impl_partial_default_value();

                    if !res.no_value {
                        let key = field.get_key();
                        let value = res.value;

                        rows.push(quote! {
                            #key: #value,
                        });
                    }

                    if res.requires_internal {
                        requires_internal = true;
                    }
                }

                // Do not implement method
                if rows.is_empty() {
                    return quote! {};
                }

                let default_row = ImplResult::impl_struct_default(rows.len() != fields.len());

                quote! {
                    Ok(Some(Self {
                        #(#rows)*
                        #default_row
                    }))
                }
            }
            ContainerInner::UnnamedStruct { fields } => {
                let mut rows = vec![];
                let mut all_none = true;

                for field in fields {
                    let res = field.impl_partial_default_value();

                    if res.no_value {
                        rows.push(quote! { None });
                    } else {
                        all_none = false;
                        let value = res.value;

                        rows.push(quote! {
                            #value
                        });
                    }

                    if res.requires_internal {
                        requires_internal = true;
                    }
                }

                // Do not implement method
                if all_none {
                    return quote! {};
                }

                quote! {
                    Ok(Some(Self(
                        #(#rows),*
                    )))
                }
            }
            ContainerInner::Enum { variants } | ContainerInner::UnitEnum { variants } => {
                let default_variants = variants
                    .iter()
                    .filter(|v| v.is_default())
                    .collect::<Vec<_>>();

                if default_variants.len() > 1 {
                    panic!("Only 1 variant may be marked as default.");
                }

                match default_variants.first() {
                    Some(default_variant) => {
                        let res = default_variant.impl_partial_default_value();

                        if res.requires_internal {
                            requires_internal = true;
                        }

                        if res.no_value {
                            quote! {
                                Ok(None)
                            }
                        } else {
                            let value = res.value;

                            quote! {
                                Ok(Some(Self::#value))
                            }
                        }
                    }
                    None => quote! {
                        Ok(None)
                    },
                }
            }
        };

        let internal = ImplResult::impl_use_internal(requires_internal);

        quote! {
            fn default_values(context: &Self::Context) -> std::result::Result<Option<Self>, schematic::ConfigError> {
                #internal
                #inner
            }
        }
    }

    #[cfg(not(feature = "env"))]
    pub fn impl_partial_env_values(&self) -> TokenStream {
        quote! {}
    }

    #[cfg(feature = "env")]
    pub fn impl_partial_env_values(&self) -> TokenStream {
        let inner = match &self.inner {
            ContainerInner::NamedStruct { fields } | ContainerInner::UnnamedStruct { fields } => {
                let mut rows = vec![];

                for field in fields {
                    let res = field.impl_partial_env_value();

                    if !res.no_value {
                        let key = field.get_key();
                        let value = res.value;

                        rows.push(quote! {
                            partial.#key = #value;
                        });
                    }
                }

                // Do not implement method
                if rows.is_empty() {
                    return quote! {};
                }

                quote! {
                    #(#rows)*
                }
            }

            // Enums don't support env vars
            _ => return quote! {},
        };

        let internal = ImplResult::impl_use_internal(true);

        let prefix_fallback = if let Some(env_prefix) = &self.args.env_prefix {
            if env_prefix.is_empty() {
                panic!("Attribute `env_prefix` cannot be empty.");
            }

            quote! { prefix.or_else(Some(#env_prefix)) }
        } else {
            quote! { prefix }
        };

        quote! {
            fn env_values_with_prefix(prefix: Option<&str>) -> std::result::Result<Option<Self>, schematic::ConfigError> {
                #internal

                let mut env = EnvManager::new(#prefix_fallback);
                let mut partial = Self::default();

                #inner

                Ok(if env.is_empty() {
                    None
                } else {
                    Some(partial)
                })
            }
        }
    }

    #[cfg(not(feature = "extends"))]
    pub fn impl_partial_extends_from(&self) -> TokenStream {
        quote! {}
    }

    #[cfg(feature = "extends")]
    pub fn impl_partial_extends_from(&self) -> TokenStream {
        if let ContainerInner::NamedStruct { fields } = &self.inner {
            let mut names = vec![];
            let mut inner = quote! { None };

            for field in fields {
                if field.is_extendable() {
                    names.push(field.get_name_original().to_string());

                    let res = field.impl_partial_extends_from();

                    if !res.no_value {
                        inner = res.value;
                    }
                }
            }

            if names.len() > 1 {
                panic!(
                    "Only 1 setting may use `extend`, found: {}",
                    names.join(", ")
                );
            }

            quote! {
                fn extends_from(&self) -> Option<schematic::ExtendsFrom> {
                    #inner
                }
            }
        } else {
            panic!("Only named structs can use `extend` settings.");
        }
    }

    pub fn impl_partial_merge(&self) -> TokenStream {
        match &self.inner {
            ContainerInner::NamedStruct { fields } | ContainerInner::UnnamedStruct { fields } => {
                let mut statements = vec![];

                for field in fields {
                    let res = field.impl_partial_merge();

                    if !res.no_value {
                        statements.push(res.value);
                    }
                }

                if statements.is_empty() {
                    return quote! {};
                }

                let internal = ImplResult::impl_use_internal(true);

                quote! {
                    fn merge(
                        &mut self,
                        context: &Self::Context,
                        mut next: Self,
                    ) -> std::result::Result<(), schematic::ConfigError> {
                        #internal

                        MergeManager::new(context)
                        #(#statements)*;

                        Ok(())
                    }
                }
            }
            ContainerInner::Enum { variants } | ContainerInner::UnitEnum { variants } => {
                let mut statements = vec![];

                for variant in variants {
                    let res = variant.impl_partial_merge();

                    if !res.no_value {
                        statements.push(res.value);
                    }
                }

                let inner = if statements.is_empty() {
                    quote! {
                        *self = next;
                    }
                } else {
                    quote! {
                        match self {
                            #(#statements)*
                            _ => {
                                *self = next;
                            }
                        };
                    }
                };

                quote! {
                    fn merge(
                        &mut self,
                        context: &Self::Context,
                        mut next: Self,
                    ) -> std::result::Result<(), schematic::ConfigError> {
                        #inner
                        Ok(())
                    }
                }
            }
        }
    }
}

impl ToTokens for Container {
    fn to_tokens(&self, _tokens: &mut TokenStream) {
        // TODO
    }
}

#[derive(Debug)]
pub enum ContainerInner {
    NamedStruct { fields: Vec<Field> },
    UnnamedStruct { fields: Vec<Field> },
    Enum { variants: Vec<Variant> },
    UnitEnum { variants: Vec<Variant> },
}

impl ContainerInner {
    pub fn get_fields(&self) -> Vec<&Field> {
        match self {
            Self::NamedStruct { fields } | Self::UnnamedStruct { fields } => {
                fields.iter().collect()
            }
            _ => vec![],
        }
    }

    pub fn get_variants(&self) -> Vec<&Variant> {
        match self {
            Self::Enum { variants } | Self::UnitEnum { variants } => variants.iter().collect(),
            _ => vec![],
        }
    }
}
