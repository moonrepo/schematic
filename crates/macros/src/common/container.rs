use crate::common::{Field, Variant};
use crate::utils::map_option_argument_quote;
use proc_macro2::TokenStream;
use quote::quote;
use syn::Fields;

pub enum Container<'l> {
    NamedStruct { fields: Vec<Field<'l>> },
    Enum { variants: Vec<Variant<'l>> },
}

impl<'l> Container<'l> {
    pub fn has_nested(&self) -> bool {
        match self {
            Self::NamedStruct { fields, .. } => fields.iter().any(|v| v.is_nested()),
            Self::Enum { variants } => variants.iter().any(|v| v.is_nested()),
        }
    }

    pub fn generate_schema(&self, description: Option<String>) -> TokenStream {
        let description = if let Some(comment) = description {
            quote! {
                schema.set_description(#comment);
            }
        } else {
            quote! {}
        };

        match self {
            Self::NamedStruct { fields, .. } => {
                let schema_types = fields
                    .iter()
                    .filter_map(|f| {
                        if f.is_excluded() {
                            None
                        } else {
                            Some(f.generate_schema_type())
                        }
                    })
                    .collect::<Vec<_>>();

                quote! {
                    #description
                    schema.structure(StructType::new([
                        #(#schema_types),*
                    ]))
                }
            }
            Self::Enum { variants } => {
                let is_all_unit_enum = variants
                    .iter()
                    .all(|v| matches!(v.value.fields, Fields::Unit));
                let mut default_index = None;

                let variants_types = variants
                    .iter()
                    .enumerate()
                    .filter_map(|(i, v)| {
                        if v.is_default() {
                            default_index = Some(i);
                        }

                        if v.is_excluded() {
                            None
                        } else {
                            Some(v.generate_schema_type(is_all_unit_enum))
                        }
                    })
                    .collect::<Vec<_>>();

                let default_index = map_option_argument_quote(default_index);

                if is_all_unit_enum {
                    quote! {
                        #description
                        schema.enumerable(EnumType::from_macro(
                            [
                                #(#variants_types),*
                            ],
                            #default_index,
                        ))
                    }
                } else {
                    quote! {
                        #description
                        schema.union(UnionType::from_macro(
                            [
                                #(#variants_types),*
                            ],
                            #default_index,
                        ))
                    }
                }
            }
        }
    }
}
