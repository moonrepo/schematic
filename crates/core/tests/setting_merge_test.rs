mod utils;

use schematic_core::container::Container;
use starbase_sandbox::assert_snapshot;
use syn::parse_quote;
use utils::pretty;

mod setting_merge {
    use super::*;

    mod named_struct {
        use super::*;

        #[test]
        fn accepts_func_ref() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example {
                    #[setting(merge = func_ref)]
                    a: String,
                }
            });
            let field = container.inner.get_fields()[0];

            assert!(field.args.merge.is_some());
        }

        #[test]
        fn accepts_string() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example {
                    #[setting(merge = "func_ref")]
                    a: String,
                }
            });
            let field = container.inner.get_fields()[0];

            assert!(field.args.merge.is_some());
        }

        #[test]
        #[should_panic(expected = "UnexpectedType")]
        fn errors_invalid_type() {
            Container::from(parse_quote! {
                #[derive(Config)]
                struct Example {
                    #[setting(merge = 123)]
                    a: String,
                }
            });
        }

        #[test]
        fn supports_standard() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example {
                    a: bool,
                    b: usize,
                    c: String,
                    d: i16,
                    e: Option<String>,
                    f: Vec<String>,
                    g: Option<HashMap<u8, String>>,
                }
            });

            assert_snapshot!(pretty(container.impl_partial_merge()));
        }

        #[test]
        fn supports_nested() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example {
                    #[setting(nested)]
                    a: NestedConfig,
                    #[setting(nested = CustomConfig)]
                    b: CustomConfig,
                    #[setting(nested)]
                    a: Option<NestedConfig>,
                    #[setting(nested = CustomConfig)]
                    b: Arc<CustomConfig>,
                }
            });

            assert_snapshot!(pretty(container.impl_partial_merge()));
        }

        #[test]
        fn supports_nested_collections() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example {
                    #[setting(nested, merge = append_vec)]
                    a: Vec<NestedConfig>,
                    #[setting(nested = CustomConfig, merge = merge_hashmap)]
                    b: HashMap<String, CustomConfig>,
                    #[setting(nested, merge = merge_btreeset)]
                    a: Option<BTreeSet<NestedConfig>>,
                }
            });

            assert_snapshot!(pretty(container.impl_partial_merge()));
        }

        #[test]
        fn supports_func() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example {
                    a: bool,
                    b: usize,
                    #[setting(merge = discard)]
                    c: String,
                    #[setting(merge = "preserve")]
                    d: i16,
                    e: Option<String>,
                    #[setting(merge = append_vec)]
                    f: Vec<String>,
                    #[setting(merge = merge_hashmap)]
                    g: Option<HashMap<u8, String>>,
                }
            });

            assert_snapshot!(pretty(container.impl_partial_merge()));
        }

        #[test]
        #[should_panic(
            expected = "Nested configs do not support `merge` unless wrapped in a collection."
        )]
        fn errors_if_nested_has_merge_attr() {
            Container::from(parse_quote! {
                #[derive(Config)]
                struct Example {
                    #[setting(nested, merge = append_vec)]
                    a: NestedConfig,
                }
            })
            .impl_partial_merge();
        }

        #[test]
        #[should_panic(expected = "Collections with nested configs must manually define `merge`.")]
        fn errors_if_collection_doesnt_have_merge_attr() {
            Container::from(parse_quote! {
                #[derive(Config)]
                struct Example {
                    #[setting(nested)]
                    a: Vec<NestedConfig>,
                }
            })
            .impl_partial_merge();
        }
    }

    mod unnamed_struct {
        use super::*;

        #[test]
        fn accepts_func_ref() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example(
                    #[setting(merge = func_ref)]
                    String,
                );
            });
            let field = container.inner.get_fields()[0];

            assert!(field.args.merge.is_some());
        }

        #[test]
        fn accepts_string() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example(
                    #[setting(merge = "func_ref")]
                    String,
                );
            });
            let field = container.inner.get_fields()[0];

            assert!(field.args.merge.is_some());
        }

        #[test]
        #[should_panic(expected = "UnexpectedType")]
        fn errors_invalid_type() {
            Container::from(parse_quote! {
                #[derive(Config)]
                struct Example(
                    #[setting(merge = 123)]
                    String,
                );
            });
        }

        #[test]
        fn supports_standard() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example(
                    bool,
                    usize,
                    String,
                    i16,
                    Option<String>,
                    Vec<String>,
                    Option<HashMap<u8, String>>,
                );
            });

            assert_snapshot!(pretty(container.impl_partial_merge()));
        }

        #[test]
        fn supports_nested() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example(
                    #[setting(nested)]
                    NestedConfig,
                    #[setting(nested = CustomConfig)]
                    CustomConfig,
                    #[setting(nested)]
                    Option<NestedConfig>,
                    #[setting(nested = CustomConfig)]
                    Arc<CustomConfig>,
                );
            });

            assert_snapshot!(pretty(container.impl_partial_merge()));
        }

        #[test]
        fn supports_nested_collections() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example(
                    #[setting(nested, merge = append_vec)]
                    Vec<NestedConfig>,
                    #[setting(nested = CustomConfig, merge = merge_hashmap)]
                    HashMap<String, CustomConfig>,
                    #[setting(nested, merge = merge_btreeset)]
                    Option<BTreeSet<NestedConfig>>,
                );
            });

            assert_snapshot!(pretty(container.impl_partial_merge()));
        }

        #[test]
        fn supports_func() {
            let container = Container::from(parse_quote! {
                #[derive(Config)]
                struct Example(
                    bool,
                    usize,
                    #[setting(merge = discard)]
                    String,
                    #[setting(merge = "preserve")]
                    i16,
                    Option<String>,
                    #[setting(merge = append_vec)]
                    Vec<String>,
                    #[setting(merge = merge_hashmap)]
                    Option<HashMap<u8, String>>,
                );
            });

            assert_snapshot!(pretty(container.impl_partial_merge()));
        }

        #[test]
        #[should_panic(
            expected = "Nested configs do not support `merge` unless wrapped in a collection."
        )]
        fn errors_if_nested_has_merge_attr() {
            Container::from(parse_quote! {
                #[derive(Config)]
                struct Example(
                    #[setting(nested, merge = append_vec)]
                    NestedConfig,
                );
            })
            .impl_partial_merge();
        }

        #[test]
        #[should_panic(expected = "Collections with nested configs must manually define `merge`.")]
        fn errors_if_collection_doesnt_have_merge_attr() {
            Container::from(parse_quote! {
                #[derive(Config)]
                struct Example(
                    #[setting(nested)]
                    Vec<NestedConfig>,
                );
            })
            .impl_partial_merge();
        }
    }
}
