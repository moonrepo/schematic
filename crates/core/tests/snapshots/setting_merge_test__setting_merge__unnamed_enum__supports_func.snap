---
source: crates/core/tests/setting_merge_test.rs
expression: pretty(container.impl_partial_merge())
---
fn merge(
    &mut self,
    context: &Self::Context,
    mut next: Self,
) -> std::result::Result<(), schematic::ConfigError> {
    match self {
        Self::A(ao) => {
            if let Self::A(ai) = next {
                *self = Self::A(a(ao.to_owned(), ai, context)?.unwrap_or_default());
            } else {
                *self = next;
            }
        }
        Self::B(ao) => {
            if let Self::B(ai) = next {
                *self = Self::B(b(ao.to_owned(), ai, context)?.unwrap_or_default());
            } else {
                *self = next;
            }
        }
        Self::C(ao, bo) => {
            if let Self::C(ai, bi) = next {
                if let Some((ao, bo)) = c(
                    (ao.to_owned(), bo.to_owned()),
                    (ai, bi),
                    context,
                )? {
                    *self = Self::C(ao, bo);
                } else {
                    *self = Self::C(Default::default(), Default::default());
                }
            } else {
                *self = next;
            }
        }
        Self::D(ao, bo) => {
            if let Self::D(ai, bi) = next {
                if let Some((ao, bo)) = d(
                    (ao.to_owned(), bo.to_owned()),
                    (ai, bi),
                    context,
                )? {
                    *self = Self::D(ao, bo);
                } else {
                    *self = Self::D(Default::default(), Default::default());
                }
            } else {
                *self = next;
            }
        }
        Self::E(ao) => {
            if let Self::E(ai) = next {
                *self = Self::E(e(ao.to_owned(), ai, context)?.unwrap_or_default());
            } else {
                *self = next;
            }
        }
        _ => {
            *self = next;
        }
    };
    Ok(())
}
